src/coreclr/nativeaot/Runtime/unix/unixasmmacrosloongarch64.inc

```diff
@@ -0,0 +1,328 @@
+
+.macro PROLOG_SAVE_REG_PAIR reg1, reg2, ofs
+        st.d  $r\reg1, $sp, \ofs
+        st.d  $r\reg2, $sp, \ofs + 8
+        .cfi_rel_offset  \reg1, \ofs
+        .cfi_rel_offset  \reg2, \ofs + 8
+        .ifc  \reg1, $fp
+        ori  $fp, $sp, 0
+        .cfi_def_cfa_register  22
+        .endif
+.endm
+
+.macro PROLOG_SAVE_REG_PAIR_INDEXED reg1, reg2, ofs
+        st.d  $r\reg1, $sp, \ofs
+        st.d  $r\reg2, $sp, \ofs + 8
+        addi.d  $sp, $sp, \ofs
+        .cfi_adjust_cfa_offset  -\ofs
+        .cfi_rel_offset  \reg1, 0
+        .cfi_rel_offset  \reg2, 8
+        .ifc  \reg1, $fp
+        ori  $fp, $sp, 0
+        .cfi_def_cfa_register  $fp
+        .endif
+.endm
+
+.macro PROLOG_SAVE_REG_PAIR_NO_FP_INDEXED reg1, reg2, ofs
+        st.d  $r\reg1, $sp, \ofs
+        st.d  $r\reg2, $sp, \ofs + 8
+        addi.d  $sp, $sp, \ofs
+        .cfi_adjust_cfa_offset  -\ofs
+        .cfi_rel_offset  \reg1, 0
+        .cfi_rel_offset  \reg2, 8
+.endm
+
+
+.macro EPILOG_RESTORE_REG reg, ofs
+        ld.d  $r\reg, $sp, \ofs
+        .cfi_restore  \reg
+.endm
+
+.macro EPILOG_RESTORE_REG_PAIR reg1, reg2, ofs
+        ld.d  $r\reg1, $sp, \ofs
+        ld.d  $r\reg2, $sp, \ofs + 8
+        .cfi_restore  \reg1
+        .cfi_restore  \reg2
+.endm
+
+.macro EPILOG_RESTORE_REG_PAIR_INDEXED reg1, reg2, ofs
+        ld.d  $r\reg1, $sp, 0
+        ld.d  $r\reg2, $sp, 8
+        addi.d  $sp, $sp, \ofs
+        .cfi_restore  \reg1
+        .cfi_restore  \reg2
+        .cfi_adjust_cfa_offset  -\ofs
+.endm
+
+.macro EPILOG_RETURN
+        jirl  $r0, $ra, 0
+.endm
+
+.macro EMIT_BREAKPOINT
+        break  0
+.endm
+
+.macro EPILOG_BRANCH_REG reg
+
+        jirl  $r0, \reg, 0
+
+.endm
+
+// Loads the address of a thread-local variable into the target register,
+// which cannot be a0. Preserves all other registers.
+.macro INLINE_GET_TLS_VAR target, var
+    .ifc  \target, $a0
+        .error "target cannot be a0"
+    .endif
+
+    st.d  $a0, $sp, -0x10
+    st.d  $ra, $sp, -0x8
+    addi.d  $sp, $sp, -16
+
+    // This sequence of instructions is recognized and potentially patched
+    // by the linker (GD->IE/LE relaxation).
+    //la.local  $a0, \var                      //
+    //ld.d  \target, $a0, 0                    //
+    //.tlsdesccall \var                        //TODO-LOONGARCH64
+    la.tls.ie   $a0, \var
+    //jirl  $ra, \target, 0
+    // End of the sequence
+
+    ori  \target, $tp, 0
+    add.d  \target, \target, $a0
+
+    ld.d  $a0, $sp, 0
+    ld.d  $ra, $sp, 8
+    addi.d $sp, $sp, 16
+.endm
+
+// Inlined version of RhpGetThread. Target cannot be a0.
+.macro INLINE_GETTHREAD target
+    INLINE_GET_TLS_VAR  \target, C_FUNC(tls_CurrentThread)
+.endm
+
+// Do not use these ETLS macros in functions that already create a stack frame.
+// Creating two stack frames in one function can confuse the unwinder/debugger
+
+.macro GETTHREAD_ETLS_1
+    PROLOG_SAVE_REG_PAIR_INDEXED   22, 1, -32          // ;; Push down stack pointer and store FP and RA
```

Comment:
After rewrite `PROLOG_SAVE_REG_PAIR_INDEXED`, all places should be positive 32,

Reply:
Got it.
I've rewritten `PROLOG_SAVE_REG_PAIR_INDEXED` and `PROLOG_SAVE_REG_PAIR_NO_FP_INDEXED`, and added atomic in the latest commit.
Thanks a lot!